functorch/_src/aot_autograd.py

```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
 #
 # TODO: Provide a faster version of this that assumes flat arguments
 # (so no pytree necessary)
-def detach_and_functionalize_pure(f, preserve_requires_grad=True):
+def run_functionalized_fw_and_collect_metadata(f):
+    def to_fun(t):
+        if isinstance(t, Tensor):
+            return torch._to_functional_tensor(t, mirror_autograd_meta=True)
+        else:
+            return t
+
+    def from_fun(t):
+        if not isinstance(t, Tensor) or not torch._is_functional_tensor(t):
+            return t
+        torch._sync(t)
+        return torch._from_functional_tensor(t)
+
     @wraps(f)
-    def inner(*args, **kwargs):
-        def to_fun(t):
-            if isinstance(t, Tensor):
-                r = torch._to_functional_tensor(t)
-                # NB: r is a leaf; it has no grad_fn relating
-                # it to t.  If t has autograd metadata, that
-                # metadata was preserved *inside* the r wrapper
-                if preserve_requires_grad:
-                    r.requires_grad = t.requires_grad
-                return r
-            else:
-                return t
+    def inner(*args):
+        # This function is meant to be run with the forward, which expects a flat list of tensor/symint/other args.
+        assert all(isinstance(a, torch.Tensor) or type(a) in KNOWN_TYPES for a in args)
 
-        f_args, f_kwargs = pytree.tree_map(to_fun, (args, kwargs))
+        collect_mutated_input_info: List[MutationType] = []
+        collect_requires_grad_out_info: List[bool] = []
+        collect_aliased_output_info: List[Optional[Union[int, torch.Tensor]]] = []
+        collect_metadata_mutation_input_info: List[Tensor] = []
+
+        f_args = pytree.tree_map(to_fun, args)
 
         torch._enable_functionalization(reapply_views=True)
         try:
-            outs = f(*f_args, **f_kwargs)
+            outs = f(*f_args)
```

Comment:
Call this `f_outs` not outs. I don't know why I didn't name it like that originally.